scipy/stats/tests/test_axis_nan_policy.py

```diff
@@ -347,6 +347,41 @@ def unpacker(res):
                 assert_equal(res1dc, res1da)
 
 
+@pytest.mark.parametrize(("hypotest", "args", "kwds", "n_samples", "n_outputs",
+                          "paired", "unpacker"), axis_nan_policy_cases)
+@pytest.mark.parametrize("nan_policy", ("omit", "propagate"))
+@pytest.mark.parametrize("axis", (None, 0, -1, (0, 2), (1, -1), (3, 1, 2, 0)))
+def test_keepdims(hypotest, args, kwds, n_samples, n_outputs, paired, unpacker,
+                  axis, nan_policy):
```

Comment:
This test might be fine (I haven't read carefully yet). I like that it checks something about `axis` `None`, negative, a tuple, and an out of order tuple. Seems like all the major cases.

On one hand, I say I like to write tests that suggest correct functionality for all possible inputs without knowledge of how the code works. This leads to very exhaustive tests. I tried to do that in the initial `_axis_nan_policy_test`. But as we expand the decorator, I don't think it's necessary. That's why we wrote separate, simpler tests when we added `axis` tuple and masked array support to the decorator rather than adding an additional layer of `parametrize`ation to the `test_axis_nan_policy_full`/`test_axis_nan_policy_fast`. It would be ok with me if you wanted to do that here rather than testing for all hypotests. I think we just need to hit these major points:
* single-output and multi-output functions (e.g. a function like `gmean` and a function like `mannwhitneyu`)
* Axis negative, positive, `None`, and tuple (as you have above)
* 1D with no NaNs, 1D with NaN propagation, zero size output, natively vectorized function with multidimensional output, and function vectorized with the decorator (either because it doesn't have native vectorization or because we're handling the vectorization because the input has NaNs or masked values)

Does that make sense? Do we have those cases here? Are there others you think are important?

Reply:
> Does that make sense? Do we have those cases here? Are there others you think are important?

You are right. On my machine, `test_axis_nan_policy` already takes a minute to run (1/3 the time taken to test the entire module). I will remove parameterization and we should probably start removing some `test_mstats` tests that overlap with that of `test_axis_nan_policy` and also the code from stats functions that call the mstats module. (I don't think we need those calls anymore)

Reply:
It takes <4s to run `test_mstats_basic.py` and `test_extras.py`. I'd prefer to just remove the test files if and when we remove `mstats` altogether.

Reply:
I have pushed a test that only tests `gmean` and `mannwhitneyu` with all the cases you outlined in your comment.

> * natively vectorized function with multidimensional output, and function vectorized with the decorator

I am sorry, I don't follow this. How should I natively vectorize since hypotests aren't elementwise operations?

Reply:
No, that's not what I meant. There were five items in that single bullet:

- 1D with no NaNs,
- 1D with NaN propagation
- zero size output
- natively vectorized function with multidimensional output
- and function vectorized with the decorator (either because it doesn't have native vectorization or because we're handling the vectorization because the input has NaNs or masked values)

Each of these items corresponds with one of five important return statements in the decorator. I wanted to make sure all of these relevant return statements (and the code paths to get to them) are tested.